### PR TITLE
[Download] Follow redirects between Hub hosts when resolving files

### LINKED ISSUE #4637: snapshot_download fails with misleading LocalEntryNotFoundError when HF_ENDPOINT mirror 308-redirects to the Hub (x-repo-commit missing); relative-redirect follower drops query string
## Summary

`snapshot_download()` fails with a misleading `LocalEntryNotFoundError` ("check your connection") whenever `HF_ENDPOINT` points to a mirror that answers with a **308 redirect back to huggingface.co** (e.g. hf-mirror.com, which has been redirecting to the main Hub since ~2026). The real failure is that the response after following the redirect **does not include the `x-repo-commit` header**, so `_get_metadata_or_catch_error` raises `FileMetadataError: "Distant resource does not seem to be on huggingface.co"`, which `_raise_on_head_call_error` then misclassifies as a connectivity problem.

I believe there are two genuine upstream defects here:

1. **`_httpx_follow_relative_redirects_with_backoff` drops the query string** when following relative redirects (`huggingface_hub/utils/_http.py:691`): it rebuilds the URL with `urlparse(url)._replace(path=parsed_target.path)` only. HF's xet storage redirects (`307 /api/resolve-cache/...?etag=...`) carry important query params; stripping them changes the response (and whether `x-repo-commit` is present).
2. **`_raise_on_head_call_error` misattributes `FileMetadataError` to network issues**, making this extremely hard to diagnose (the error text says "check your connection" while the network is perfectly fine).

## Environment

- huggingface_hub 1.24.0
- faster-whisper 1.2.1 (calls `snapshot_download`)
- Python 3.11.15, Windows 10

## Repro

```python
# With HF_ENDPOINT set to a mirror that 308-redirects to huggingface.co:
#   set HF_ENDPOINT=https://hf-mirror.com   (hf-mirror.com now 308s to huggingface.co)
from huggingface_hub import snapshot_download
snapshot_download("Systran/faster-whisper-base")
# -> LocalEntryNotFoundError: An error happened while trying to locate the file on the Hub and
#    we cannot find the requested files in the local cache. Please check your connection...
```

Root cause chain (verified):

1. `hf_hub_url(..., endpoint=HF_ENDPOINT)` builds `https://hf-mirror.com/Systran/faster-whisper-base/resolve/main/config.json`
2. `hf-mirror.com` answers **308 Permanent Redirect** → `https://huggingface.co/Systran/faster-whisper-base/resolve/main/config.json`
3. `get_hf_file_metadata` follows the redirect, but the final response has **no `x-repo-commit` header** → `commit_hash is None`
4. `_get_metadata_or_catch_error` raises `FileMetadataError: "Distant resource does not seem to be on huggingface.co"` (file_download.py:1735)
5. `_raise_on_head_call_error` (file_download.py:1867) converts it to `LocalEntryNotFoundError` with the generic "connection" message.

## Evidence

Direct HEAD to the Hub works fine:

```
$ curl -sI https://huggingface.co/Systran/faster-whisper-base/resolve/main/config.json
HTTP/1.1 307 Temporary Redirect
Location: /api/resolve-cache/models/Systran/faster-whisper-base/<commit>/config.json?%2FSystran%2Ffaster-whisper-base%2Fresolve%2Fmain%2Fconfig.json=&etag=...

$ python -c "import httpx; r=httpx.head('https://huggingface.co/Systran/faster-whisper-base/resolve/main/config.json', follow_redirects=True); print(r.status_code, r.headers.get('x-repo-commit'))"
200 ebe41f70d5b6dfa9166e2c581c45c9c0cfc57b66
```

But through the mirror endpoint the header is missing:

```
$ curl -sI https://hf-mirror.com/Systran/faster-whisper-base/resolve/main/config.json
HTTP/1.1 308 Permanent Redirect
Location: https://huggingface.co/Systran/faster-whisper-base/resolve/main/config.json
```

Monkeypatching `get_hf_file_metadata` inside `snapshot_download` shows every file resolves to the mirror and comes back with `x-repo-commit: None`:

```
!!! commit=None: config.json    status: 308 final: https://hf-mirror.com/Systran/faster-whisper-base/resolve/<commit>/config.json
!!! commit=None: .gitattributes status: 308 final: https://hf-mirror.com/Systran/faster-whisper-base/resolve/<commit>/.gitattributes
...
总体FAIL: LocalEntryNotFoundError
```

Also note the relative-redirect follower in `_httpx_follow_relative_redirects_with_backoff` (utils/_http.py:691):

```python
if parsed_target.netloc == "":
    # Relative redirect -> update URL and retry
    url = urlparse(url)._replace(path=parsed_target.path).geturl()
    continue
```

…drops `parsed_target.query`. For HF xet redirects (307 → `/api/resolve-cache/...?etag=...`) the query carries the etag; the response to the query-less URL behaves differently (and lacks `x-repo-commit` in the mirror scenario above).

## Suggested fix

1. Preserve the query when following relative redirects:
   ```python
   url = urlparse(url)._replace(path=parsed_target.path, query=parsed_target.query).geturl()
   ```
2. In `_raise_on_head_call_error`, include the underlying exception type/message in the user-facing error (or at least don't claim "check your connection" for `FileMetadataError`/missing-header cases).

## Workaround

Unset `HF_ENDPOINT` (or point it at a mirror that does not redirect to the Hub). With the default endpoint everything works.

### PR DESCRIPTION
**Pull request by hanouticelina**
Fixes #4637.

## Summary

- When resolving a file, the client only followed redirects with a relative `Location`. An endpoint answering with an absolute redirect to another Hub host (e.g. a mirror set as `HF_ENDPOINT` redirecting to `huggingface.co`) was never followed: the client kept the raw `308`, found no `X-Repo-Commit` header on it, and failed with a misleading `LocalEntryNotFoundError: ... check your connection`.
- New rule: follow a redirect when the target stays on the same host or on a known Hub host (`constants.HF_URL_HOSTS`, i.e. `huggingface.co`, `hf.co`, staging, and the host of the configured `HF_ENDPOINT`). Stop at any other host: for LFS files the metadata lives on the redirect response itself, and the `Authorization` header must not be forwarded to a CDN.
- More accurate error when metadata is genuinely missing: the message now names the URL and the missing header and points at `HF_ENDPOINT`, instead of blaming the connection.

---

> [!NOTE]
> **Medium Risk**
> Changes core HEAD/resolve and auth-forwarding behavior for downloads; logic is tightened around trusted Hub hosts but misconfigured `HF_URL_HOSTS` could affect redirects or token forwarding.
> 
> **Overview**
> Fixes file resolution when `HF_ENDPOINT` (or another Hub mirror) answers with an **absolute redirect** to a known Hub host (e.g. `huggingface.co`). The client previously only chased **relative** `Location` headers, so it could stop on a `308` without `X-Repo-Commit` and surface a misleading “check your connection” cache error.
> 
> **Redirect policy:** `_httpx_follow_relative_redirects_with_backoff` is replaced by `_httpx_follow_hub_redirects_with_backoff`, which follows redirects while the target stays on the **same host** or on `constants.HF_URL_HOSTS` (public Hub, `hf.co`, staging, configured endpoint). Redirects to CDNs/storage are **not** followed so LFS metadata on the redirect response is preserved and **`Authorization` is not sent off-Hub**. `_is_same_or_hub_host` is shared for HEAD chasing and for stripping auth on CDN download URLs.
> 
> **Errors:** Missing `X-Repo-Commit` now names the URL, the header, and `HF_ENDPOINT`. `FileMetadataError` during cache fallback is re-raised as `LocalEntryNotFoundError` with a configuration-focused message instead of a generic connectivity warning.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 809e4fbbf0bac07911df52b8c0f49e2402fdb110. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Review by hanouticelina — COMMENTED** [review]
✅

File: src/huggingface_hub/utils/_http.py [hunk 1]
```diff
@@ -728,18 +732,24 @@ def _httpx_follow_relative_redirects_with_backoff(
...
-        if 300 <= response.status_code <= 399:
-            parsed_target = urlparse(response.headers["Location"])
-            if parsed_target.netloc == "":
-                # Relative redirect -> update URL and retry
-                url = urlparse(url)._replace(path=parsed_target.path).geturl()
-                continue
+        if not response.has_redirect_location:
+            return response
+
+        target = urljoin(url, response.headers["Location"])
```
Comment (Wauplin) [inline src/huggingface_hub/utils/_http.py:738]: TIL (again?) `urljoin` behavior is so confusing 😵 

https://stackoverflow.com/questions/10893374/python-confusions-with-urljoin

**Review by Wauplin — APPROVED** [review]
LGTM, thank you!

File: src/huggingface_hub/utils/_http.py (hunk 1 above)
Comment (hanouticelina) [reply src/huggingface_hub/utils/_http.py:738]: lol Claude sent me to the same stackoverflow page because i didn't understand much in `urljoin` behavior

Comment (Wauplin) [reply src/huggingface_hub/utils/_http.py:738]: well 14 years ago and still top 1 google result so we are not the only ones^^

**Comment by Puneet15556** [comment]
Late follow-up on this after it landed — there's one case the host check doesn't cover.

`_is_same_or_hub_host` compares hostnames only, so a `Location` of `http://huggingface.co/...` is judged to be a Hub host. Both call sites then act on it: `_httpx_follow_hub_redirects_with_backoff` follows the redirect, and `_get_metadata_or_catch_error` keeps the `authorization` header for the download. The token can therefore go out over plaintext http.

Verified against `main`:

| request | target | treated as Hub |
| --- | --- | --- |
| `https://hf-mirror.com/...` | `https://huggingface.co/...` | yes (correct) |
| `https://hf-mirror.com/...` | `http://huggingface.co/...` | **yes** |
| `https://hf-mirror.com/...` | `http://hf-mirror.com/...` | **yes** |
| `https://hf-mirror.com/...` | `https://cdn.example.com/...` | no (correct) |

It needs a `Location` that the endpoint controls, so in practice this is a custom `HF_ENDPOINT` mirror rather than the default endpoint — I'd call it hardening rather than a live vulnerability. But since the point of the host check is to keep the token on the Hub, the downgrade seems worth closing as well. #4648 had ended up with an explicit `https` requirement for the same reason before it was superseded.

The change is small — refuse the target when it downgrades https to http, leaving endpoints deliberately configured over http untouched:

```python
parsed_url, parsed_target = urlparse(url), urlparse(target)
if parsed_url.scheme == "https" and parsed_target.scheme != "https":
    return False
```

I also wrote tests for `_httpx_follow_hub_redirects_with_backoff` and `_is_same_or_hub_host`, which currently have none: target query preserved, base query not inherited, dot-segments resolved, absolute redirect to a Hub host followed, redirect off the Hub returned as-is, downgrade refused, and the `_MAX_REDIRECTS` bound raising `TooManyRedirects`. The three downgrade assertions fail without the change. Locally `tests/test_utils_http.py` is 80 passed and `tests/test_file_download.py` 63 passed / 4 skipped.

It's ready on a branch: https://github.com/Puneet15556/huggingface_hub/tree/fix/hub-redirect-scheme-and-tests — I tried to open it as a PR but PR creation from forks appears to be restricted on this repo at the moment. Happy to open one as soon as that's possible, or a maintainer is welcome to take the branch; equally happy to send just the tests if the scheme change isn't wanted.
